circular constraint issue

### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported


### Mongoose version

7.0.0

### Node.js version

16.14.2

### MongoDB server version

4.4

### Typescript version (if applicable)

4.9.5

### Description

I get a compilation error when specifying the `Schema` inside a `connection.model` call.


### Steps to Reproduce

I have the following line of code in a typescript project:
```
const allUserDocuments: UserDocument[] = await connection
        .model("Users", UserSchema)
        .find({})
        .exec();
```

When compiling the project, I get:

```
node_modules/typescript/lib/lib.es5.d.ts:1584:11 - error TS2313: Type parameter 'P' has a circular constraint.

1584     [P in K]: T[P];
               ~

  example.ts:60:52
         const allUserDocuments: UserDocument[] = await connection
                                                          ~~~~~~~~~~
             .model("Users", UserSchema)
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    Circularity originates in type at this location.
```

If I remove the `UserSchema` from the parameters like that:

```
    const allUserDocuments: UserDocument[] = await connection
        .model("Users")
        .find({})
        .exec();
```
it works.

doing `skipLibCheck: true` also resolves the problem, but that's not sth I wanna do.

### Expected Behavior

This error should not happen.

Mongoose version: 7.0.2
Node version: 16.16.0
Typescript version: 4.9.5

I've also been getting this issue, and might be able to provide some more insight. An example of the offending code w/ schema definitions:

```
import { Schema, model, InferSchemaType } from "mongoose";

const DisbursementSchema = new Schema({
    paymentId: {
        type: String,
        required: true
    },
    paymentDate: {
        type: Date,
        required: true
    },
    buyer: {
        id: {
            type: Schema.Types.ObjectId,
            ref: "user"
        },
        username: {
            type: String
        },
        email: {
            type: String
        },
        city: {
            type: String
        },
        state: {
            type: String
        },
        deliveryZip: {
            type: String
        },
    },
    seller: {
        id: {
            type: Schema.Types.ObjectId,
            ref: "user"
        },
        username: {
            type: String
        },
        email: {
            type: String
        },
        createDate: {
            type: Date
        }
    },
    purchases: [
        ...
    ]
});

export type DisbursementModel = InferSchemaType<typeof DisbursementSchema>;

DisbursementSchema.index({ "paymentId": 1, "seller.id": 1 }, { unique: true });

export default model("disbursement", DisbursementSchema);
```

Compiling the project results in the following error, but **only when used with the --build flag**:

```
> tsc --build

node_modules/typescript/lib/lib.es5.d.ts:1584:11 - error TS2313: Type parameter 'P' has a circular constraint.

1584     [P in K]: T[P];
               ~

  src/models/accounting/Disbursement.ts:207:33
    207 export type DisbursementModel = InferSchemaType<typeof DisbursementSchema>;
                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    Circularity originates in type at this location.


Found 1 error.
```

Avoiding the --build flag fixes the issue, though _I was able to use the --build flag before I switched to automatic schema type inference._ For those that do need the perks of typescript's build flag and don't want to maintain separate schema interfaces, I can see this being an issue.

@vkarpov15 I noticed same circular error in version 7. TL;DR: it's because of `incremental=true` in `tsconfig.json`

Simple example:

`src/locality.ts`:
```ts
import { InferSchemaType, Schema } from 'mongoose';

const schema = new Schema({});

export type LocalityRaw = InferSchemaType<typeof schema>;
```
Output:
```
> tsc

node_modules/typescript/lib/lib.es5.d.ts:1584:11 - error TS2313: Type parameter 'P' has a circular constraint.

1584     [P in K]: T[P];
               ~

  src/locality.ts:5:27
    5 export type LocalityRaw = InferSchemaType<typeof schema>;
                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    Circularity originates in type at this location.


Found 1 error in node_modules/typescript/lib/lib.es5.d.ts:1584
```

`tsconfig.json`:
```json
{
    "compilerOptions": {
        "module": "commonjs",
        "target": "esnext",
        "outDir": "dist",
        "incremental": true
    },
    "include": [
        "src/**/*"
    ]
}
```

But if I change `incremental` to `false` or remove it at all, it starts working. In version `6.10.4` I didn't have such issue.

--- a/test/types/schema.test.ts
+++ b/test/types/schema.test.ts
@@ -10,7 +10,7 @@ import {
   Query,
   model,
   HydratedDocument,
-  SchemaOptions,
+  ResolveSchemaOptions,
   ObtainDocumentType,
   ObtainSchemaGeneric
 } from 'mongoose';
@@ -634,7 +634,8 @@ function gh12003() {
 
   type BaseSchemaType = InferSchemaType<typeof BaseSchema>;
 
-  expectType<'type'>({} as ObtainSchemaGeneric<typeof BaseSchema, 'TSchemaOptions'>['typeKey']);
+  type TSchemaOptions = ResolveSchemaOptions<ObtainSchemaGeneric<typeof BaseSchema, 'TSchemaOptions'>>;
+  expectType<'type'>({} as TSchemaOptions['typeKey']);
 
   expectType<{ name?: string }>({} as BaseSchemaType);
 }

--- a/types/index.d.ts
+++ b/types/index.d.ts
@@ -211,15 +211,21 @@ declare module 'mongoose' {
     TQueryHelpers = {},
     TVirtuals = {},
     TStaticMethods = {},
-    TSchemaOptions extends ResolveSchemaOptions<TSchemaOptions> = DefaultSchemaOptions,
-    DocType extends ApplySchemaOptions<ObtainDocumentType<DocType, EnforcedDocType, TSchemaOptions>, TSchemaOptions> = ApplySchemaOptions<ObtainDocumentType<any, EnforcedDocType, TSchemaOptions>, TSchemaOptions>,
+    TSchemaOptions = DefaultSchemaOptions,
+    DocType extends ApplySchemaOptions<
+      ObtainDocumentType<DocType, EnforcedDocType, ResolveSchemaOptions<TSchemaOptions>>,
+      ResolveSchemaOptions<TSchemaOptions>
+    > = ApplySchemaOptions<
+      ObtainDocumentType<any, EnforcedDocType, ResolveSchemaOptions<TSchemaOptions>>,
+      ResolveSchemaOptions<TSchemaOptions>
+    >,
     THydratedDocumentType = HydratedDocument<FlatRecord<DocType>, TVirtuals & TInstanceMethods>
   >
     extends events.EventEmitter {
     /**
      * Create a new schema
      */
-    constructor(definition?: SchemaDefinition<SchemaDefinitionType<EnforcedDocType>> | DocType, options?: SchemaOptions<FlatRecord<DocType>, TInstanceMethods, TQueryHelpers, TStaticMethods, TVirtuals, THydratedDocumentType> | TSchemaOptions);
+    constructor(definition?: SchemaDefinition<SchemaDefinitionType<EnforcedDocType>> | DocType, options?: SchemaOptions<FlatRecord<DocType>, TInstanceMethods, TQueryHelpers, TStaticMethods, TVirtuals, THydratedDocumentType> | ResolveSchemaOptions<TSchemaOptions>);
 
     /** Adds key path / schema type pairs to this schema. */
     add(obj: SchemaDefinition<SchemaDefinitionType<EnforcedDocType>> | Schema, prefix?: string): this;

--- a/types/inferschematype.d.ts
+++ b/types/inferschematype.d.ts
@@ -46,7 +46,7 @@ declare module 'mongoose' {
    * @param {TSchema} TSchema A generic of schema type instance.
    * @param {alias} alias Targeted generic alias.
    */
-   type ObtainSchemaGeneric<TSchema, alias extends 'EnforcedDocType' | 'M' | 'TInstanceMethods' | 'TQueryHelpers' | 'TVirtuals' | 'TStaticMethods' | 'TSchemaOptions' | 'DocType'> =
+  type ObtainSchemaGeneric<TSchema, alias extends 'EnforcedDocType' | 'M' | 'TInstanceMethods' | 'TQueryHelpers' | 'TVirtuals' | 'TStaticMethods' | 'TSchemaOptions' | 'DocType'> =
    TSchema extends Schema<infer EnforcedDocType, infer M, infer TInstanceMethods, infer TQueryHelpers, infer TVirtuals, infer TStaticMethods, infer TSchemaOptions, infer DocType>
      ? {
        EnforcedDocType: EnforcedDocType;
@@ -60,8 +60,7 @@ declare module 'mongoose' {
      }[alias]
      : unknown;
 
-  // Without Omit, this gives us a "Type parameter 'TSchemaOptions' has a circular constraint."
-  type ResolveSchemaOptions<T> = Omit<MergeType<DefaultSchemaOptions, T>, 'fakepropertyname'>;
+  type ResolveSchemaOptions<T> = MergeType<DefaultSchemaOptions, T>;
 
   type ApplySchemaOptions<T, O = DefaultSchemaOptions> = ResolveTimestamps<T, O>;
 

